### PR TITLE
[super_editor_markdown] Implement table parsing. (Resolves #2758)

### LINKED ISSUE #2758: [Markdown] - Add table parsing
Within our larger efforts related to tables, one thing we want to support is Markdown tables.

There's no standard spec for Markdown tables. However, there is a popular extension specification:
https://www.markdownguide.org/extended-syntax/#tables

It looks like GitHub replicates this particular extension:
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-tables

For this ticket, handle all possible parsing of this syntax. 

Sometime after this ticket we'll send this data into a new `TableNode`, but that's outside the scope of this ticket. For now, create whatever data structure makes sense and we'll rework it in a followup ticket.

Ensure that all obvious edge cases are covered with tests (Note: There is a table generator linked from the markdown guide above - that might help with assembling test tables).

Make sure you note the statement in the guide about the types of inline formatting that are allowed, but that block level formatting isn't supported in tables for Markdown.

### PR DESCRIPTION
**Pull request by angelosilvestre**
[super_editor_markdown] Implement table parsing. (Resolves #2758)

### CODE REVIEW

File: super_editor_markdown/lib/src/markdown_inline_parser.dart [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import 'package:super_editor_markdown/super_editor_markdown.dart';
+
+/// Parses inline markdown content.
+///
+/// Supports strikethrough, underline, bold, italics, underlines links and images.
+///
+/// The given [syntax] controls how the [text] is parsed, e.g., [MarkdownSyntax.normal]
+/// for strict Markdown parsing, or [MarkdownSyntax.superEditor] to use Super Editor's
+/// extended syntax.
+InlineMarkdownToDocument parseInlineMarkdown(
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_inline_parser.dart:13]: This is an existing method that I just extracted. But we also use the `isImage` property to parse images.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 2]
```diff
@@ -72,6 +65,59 @@ MutableDocument deserializeMarkdownToDocument(
...
+///
+/// Returns the parsed nodes.
+List<md.Node> parseMarkdownNodes(
+  String markdown, {
+  MarkdownSyntax syntax = MarkdownSyntax.superEditor,
+  List<md.BlockSyntax> customBlockSyntax = const [],
+  List<ElementToNodeConverter> customElementToNodeConverters = const [],
+  bool encodeHtml = false,
+}) {
+  final lines = const LineSplitter().convert(markdown);
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart:86]: The places that are splitting lines are separate global function. They don't call each other.

File: super_editor_markdown/test/table_parser_test.dart [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+import 'package:flutter_test/flutter_test.dart';
+import 'package:markdown/markdown.dart';
+import 'package:super_editor/super_editor.dart';
+import 'package:super_editor_markdown/src/table.dart';
+import 'package:super_editor_markdown/super_editor_markdown.dart';
+
+void main() {
+  group('Markdown > deserialization > tables >', () {
+    group('parses', () {
+      test('table with single column', () {
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/test/table_parser_test.dart:12]: It shouldn't be. The table syntax require the headers and the delimiters. We could have a single header table, but not a table without any headers.

File: super_editor_markdown/test/table_parser_test.dart [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+        );
+      });
+
+      test('table broken by block level element', () {
+        expect(
+          _parseMarkdownTable(
+            '''| header 1 | header 2 |
+|---|---|
+| data 1 | data 2 |
+> blockquote''',
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/test/table_parser_test.dart:452]: > Is this testing when content follows the table?

Yes, to ensure that a block level syntax stops the table.

> If so, should we have a test for a blank line after the table?

There is already one.

File: super_editor_markdown/test/table_parser_test.dart [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+                header: AttributedText('header 2'),
+              ),
+            ],
+            rows: [
+              [
+                AttributedText('data 1'),
+                AttributedText('data 2'),
+              ],
+              [
+                AttributedText('data 3'),
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/test/table_parser_test.dart:498]: The table syntax doesn't require leading and trailing pipes, so the last line is considered a table row with a single column. It's equivalent to the following:

```
| header 1 | header 2 |
|---|---|
| data 1 | data 2 |
| data 3 |
```

Only blank lines or block level syntaxes stop the table.

File: super_editor_markdown/lib/src/markdown_inline_parser.dart [hunk 6]
```diff
@@ -0,0 +1,138 @@
+import 'package:markdown/markdown.dart' as md;
+import 'package:super_editor/super_editor.dart';
+import 'package:super_editor_markdown/src/image_syntax.dart';
+import 'package:super_editor_markdown/super_editor_markdown.dart';
+
+/// Parses inline markdown content.
+///
+/// Supports strikethrough, underline, bold, italics, underlines links and images.
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_inline_parser.dart]: Added support for inline code.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 7]
```diff
@@ -72,6 +65,59 @@ MutableDocument deserializeMarkdownToDocument(
   return MutableDocument(nodes: documentNodes);
 }
 
+/// Parses the given [markdown] without converting it to a [MutableDocument].
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart]: Updated.

File: super_editor_markdown/lib/src/table.dart [hunk 8]
```diff
@@ -0,0 +1,134 @@
+import 'dart:ui';
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/table.dart:1]: We have a bit of a mix:

- tasks
- list_items
- paragraph
- image
- horizontal_rule
- blockquote
- paragraph

File: super_editor_markdown/lib/src/table.dart [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+import 'package:markdown/markdown.dart' as md;
+import 'package:super_editor/super_editor.dart';
+import 'package:super_editor_markdown/src/markdown_inline_parser.dart';
+
+/// Converts a [md.Element] representing a table into a [MarkdownTable].
+///
+/// The [md.Element] must have a `table` tag.
+///
+/// Throws an exception if the element is not a valid table structure.
+MarkdownTable convertTable(md.Element element) {
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/table.dart:13]: What about `markdownElementToTable`? I'm not sure yet if we will keep this after implementing the table node.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 10]
```diff
@@ -161,7 +207,7 @@ class _MarkdownToDocument implements md.NodeVisitor {
         _addHeader(element, level: 6);
         break;
       case 'p':
-        final inlineVisitor = _parseInline(element.textContent);
+        final inlineVisitor = parseInlineMarkdown(element.textContent, syntax: syntax, encodeHtml: _encodeHtml);
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart]: Both `_addHeader` and `_addBlockquote` internally parse the inline markdown.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 11]
```diff
@@ -406,132 +449,51 @@ class _MarkdownToDocument implements md.NodeVisitor {
...
+      text,
+      syntax: syntax,
+      encodeHtml: _encodeHtml,
+    );
   }
 
-  _InlineMarkdownToDocument _parseInline(String text) {
+  _MarkdownImage? _maybeParseBlockImage(String markdown) {
+    if (!markdown.startsWith("![")) {
+      // A block image should start with the image syntax at the beginning of the line.
```
Comment (matthew-carroll) [inline super_editor_markdown/lib/src/markdown_to_document_parsing.dart:469]: Let's not write comments in a passive tense about "should". It's not clear what "should" means in code comments. In this case you can say `// Text doesn't start with Markdown image syntax. Return.`.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 12]
```diff
@@ -406,132 +449,51 @@ class _MarkdownToDocument implements md.NodeVisitor {
...
-          md.StrikethroughSyntax(),
-          UnderlineSyntax(),
           if (syntax == MarkdownSyntax.superEditor) //
             SuperEditorImageSyntax(),
         ],
-        encodeHtml: _encodeHtml,
       ),
     );
-    final inlineVisitor = _InlineMarkdownToDocument();
+    final inlineVisitor = InlineMarkdownToDocument();
```
Comment (matthew-carroll) [inline super_editor_markdown/lib/src/markdown_to_document_parsing.dart:482]: Is this `InlineMarkdownToDocument` only being used to find images? If so, it's probably mis-named, and overly complicated.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 13]
```diff
@@ -895,6 +857,20 @@ class _HeaderWithAlignmentSyntax extends md.BlockSyntax {
...
+    required this.url,
+    this.altText,
+    this.width,
+    this.height,
+  });
+
+  final String url;
+  final String? altText;
+  final String? width;
+  final String? height;
```
Comment (matthew-carroll) [inline super_editor_markdown/lib/src/markdown_to_document_parsing.dart:871]: Should probably implement `==` and `hashCode`?

File: super_editor_markdown/lib/src/table.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_editor_markdown/lib/src/table.dart:13]: Can we define this as an extension on `Element` and call this `asTable()`?

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 14]
```diff
@@ -23,26 +26,16 @@ MutableDocument deserializeMarkdownToDocument(
   List<ElementToNodeConverter> customElementToNodeConverters = const [],
   bool encodeHtml = false,
 }) {
-  final markdownLines = const LineSplitter().convert(markdown).map<md.Line>((String l) {
-    return md.Line(l);
-  }).toList();
+  final lines = const LineSplitter().convert(markdown);
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart:29]: It's because it was previously an array of `Line`s, so I thought this way it would be more clear that this is a plain string. But I'll revert it.

File: super_editor_markdown/lib/src/table.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/table.dart:13]: Yeah, modified.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart (hunk 12 above)
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart:482]: No, it's used to find all inline syntaxes. Bold, italics, links, etc.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 15]
```diff
@@ -406,132 +449,51 @@ class _MarkdownToDocument implements md.NodeVisitor {
...
+    // ignore: unused_local_variable
+    final table = convertTable(element);
+
+    // TODO: add table node to _content.
+  }
+
   AttributedText _parseInlineText(String text) {
-    final inlineVisitor = _parseInline(text);
-    return inlineVisitor.attributedText;
+    return parseInlineMarkdown(
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart:460]: I added just to avoid needing to pass the `syntax` and `encodeHtml` to `parseInlineMarkdown` in multiple places.

I can remove it if you think it's better to call `parseInlineMarkdown` directly.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 16]
```diff
@@ -72,6 +65,59 @@ MutableDocument deserializeMarkdownToDocument(
...
+  List<md.BlockSyntax> customBlockSyntax = const [],
+  List<ElementToNodeConverter> customElementToNodeConverters = const [],
+  bool encodeHtml = false,
+}) {
+  final lines = const LineSplitter().convert(markdown);
+
+  return _parseMarkdownLines(lines);
+}
+
+List<md.Node> _parseMarkdownLines(
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart:91]: It's because we need to call it from two different places, so I was trying to avoid code duplication.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 17]
```diff
@@ -72,6 +65,59 @@ MutableDocument deserializeMarkdownToDocument(
...
+/// The given [syntax] controls how the [markdown] is parsed, e.g., [MarkdownSyntax.normal]
+/// for strict Markdown parsing, or [MarkdownSyntax.superEditor] to use Super Editor's
+/// extended syntax.
+///
+/// To add support for parsing non-standard Markdown blocks, provide [customBlockSyntax]s
+/// that parse Markdown text into [md.Element]s, and provide [customElementToNodeConverters] that
+/// turn those [md.Element]s into [DocumentNode]s.
+///
+/// Returns the parsed nodes.
+List<md.Node> parseMarkdownNodes(
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart:79]: I made this public because I'm using it in the tests, so it cannot be private.

I don't think we currently have a class where this would make sense as a static method. Should we create one?

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart [hunk 18]
```diff
@@ -72,6 +65,59 @@ MutableDocument deserializeMarkdownToDocument(
   return MutableDocument(nodes: documentNodes);
 }
 
+/// Parses the given [markdown] to a tree representing the HTML structure of the markdown.
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart:68]: Yeah, this is about the html structure. It has `thead`, `tbody`, `tr`, `td`,  `th`, which are HTML concepts.

File: super_editor_markdown/lib/src/markdown_inline_parser.dart [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+/// obtain an [AttributedText] that represents the inline
+/// styles within the given text.
+///
+/// Apply [InlineMarkdownToDocument] to an [md.Element] whose
+/// content is an image tag to obtain image data.
+///
+/// [InlineMarkdownToDocument] does not support parsing text
+/// that contains image tags. If any non-image text is found,
+/// the content is treated as styled text.
+class InlineMarkdownToDocument implements md.NodeVisitor {
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_inline_parser.dart:51]: I think now we can make `InlineMarkdownToDocument` private again, and define a new `NodeVisitor` specific to handling images, also private.

Updated it.

File: super_editor_markdown/lib/src/markdown_inline_parser.dart [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+///
+/// Supports strikethrough, underline, bold, italics, code, links and images.
+///
+/// The given [syntax] controls how the [text] is parsed, e.g., [MarkdownSyntax.normal]
+/// for strict Markdown parsing, or [MarkdownSyntax.superEditor] to use Super Editor's
+/// extended syntax.
+AttributedText parseInlineMarkdown(
+  String text, {
+  MarkdownSyntax syntax = MarkdownSyntax.superEditor,
+  bool encodeHtml = false,
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_inline_parser.dart:16]: Added.

File: super_editor_markdown/lib/src/markdown_to_document_parsing.dart (hunk 17 above)
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart:79]: Yeah, I think so. Because I don't think any user who finds a global method in this package called `parseMarkdownNodes` is going to expect a list of `Node`s from the Dart `markdown` package.

Also, if it's **truly** only public for testing, then please annotate with a visible for testing.

[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor_markdown/lib/src/markdown_to_document_parsing.dart:80]: As discussed, marking this method as deprecated, because it is needed only while we don't have the table node.

**Comment by matthew-carroll** [comment]
@angelosilvestre now that we've merged the table node, do you want to go ahead and try using that directly, and see if it removes the need for the intermediate data structure?

**Review by matthew-carroll — APPROVED** [review]
